### PR TITLE
Add file listing to package details page

### PR DESCRIPTION
**Pull request by Daniel15**
Whenever I want to use a package via [unpkg](https://unpkg.com/) or a similar CDN, I end up doing `yarn add foo` locally, then inspecting the `node_modules` directory to find the name of the file to use. unpkg provide directory listings (eg. https://unpkg.com/react@15.5.4/), but then I have to remember to switch between the Yarn details page and unpkg.

Helpfully, unpkg also offers the directory listing as JSON (eg. https://unpkg.com/react@15.5.4/?json). I've used this to add a directory listing feature to the Yarn package details page. This pull requests add a "Browse Files" link next to the "Try in RunKit" link. When clicked, it shows a file tree for the package:

Clicking any file name will open the file via unpkg.

The icons aren't very nice, they're just royalty-free icons I found online. @Haroenv - Would it be possible to get some file/folder icons in the same style as the icons in the sidebar? It'd be nice to have custom file icons for most popular file types (JS, JSON, Markdown) too.

cc @Haroenv

### CODE REVIEW

**Comment by Haroenv** [comment]
This is an amazing idea!

File: js/src/lib/Details/index.js [hunk 1]
```diff
@@ -179,6 +171,54 @@ class Details extends React.Component {
...
+        humanDependents={this.state.humanDependents}
+        stargazers={
+          this.state.github ? this.state.github.stargazers_count : false
+        }
+        onOpenFileBrowser={this._openFileBrowser}
+      />
+    );
+  }
+
+  _renderFileBrowser() {
```
Comment (Haroenv) [inline js/src/lib/Details/index.js]: Why are all of them prefixed with an underscore?

File: js/src/lib/Details/FileBrowser.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+      // Special case for root - Only render the contents, not the outer
+      // wrapper.
+      return this._renderDirContents();
+    }
+    return (
+      <li key={dir.path}>
+        <a
+          className="details-files__dirname"
+          href={url}
+          target="_blank"
```
Comment (Haroenv) [inline js/src/lib/Details/FileBrowser.js]: No referrer and noopener

File: js/src/lib/Details/Aside.js [hunk 3]
```diff
@@ -41,6 +43,10 @@ const Aside = (
         >
           {window.i18n.detail.try_in_runkit}
         </a>
+        {' \xB7 '}
```
Comment (Haroenv) [inline js/src/lib/Details/Aside.js]: This seems a bit magic. What is it?

File: js/src/lib/Details/Aside.js [hunk 4]
```diff
@@ -23,6 +23,8 @@ const Aside = (
     humanDependents,
     dependencies,
     devDependencies,
+
```
Comment (Haroenv) [inline js/src/lib/Details/Aside.js]: Is there a reason for this?

File: assets/detail/ico-folder.svg [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<svg version="1.1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 58 58">
+  <path
+    style="fill:#EFCE4A"
+    d="M46.324,52.5H1.565c-1.03,0-1.779-0.978-1.51-1.973l10.166-27.871
+    c0.184-0.682,0.803-1.156,1.51-1.156H56.49c1.03,0,1.51,0.984,1.51,1.973L47.834,51.344C47.65,52.026,47.031,52.5,46.324,52.5z"
+  />
+  <path
+    style="fill:#EBBA16"
```
Comment (Haroenv) [inline assets/detail/ico-folder.svg]: The colours might not really fit here. Does @Shipow have a good idea for thzt

File: _data/i18n/en.yml [hunk 6]
```diff
@@ -306,11 +306,16 @@ script:
...
     last_commit: Last commit
+    loading: Loading...
     use_it: Use it
     try_in_runkit: Try in RunKit
+    back_to_details: Back to Details
+    browse_files: Browse Files
     contributors: Contributors
     display_full_readme: Display full readme
     display_full_changelog: Display full changelog
+    files_error: 'Could not load file listing: {error}'
```
Comment (Haroenv) [inline _data/i18n/en.yml]: What will that error look like? Will it be translated

File: js/src/lib/Details/index.js (hunk 1 above)
Comment (Daniel15) [inline js/src/lib/Details/index.js]: At Facebook we use an underscore prefix for all private methods 😃

File: js/src/lib/Details/Aside.js (hunk 4 above)
Comment (Daniel15) [inline js/src/lib/Details/Aside.js]: Oops, no, it was accidental.

File: js/src/lib/Details/Aside.js (hunk 3 above)
Comment (Daniel15) [inline js/src/lib/Details/Aside.js]: It's a dot: &middot; I could use `&middot;` instead.

File: _data/i18n/en.yml (hunk 6 above)
Comment (Daniel15) [inline _data/i18n/en.yml]: It's whatever error message is returned by the server (if a HTTP error occurs), or the error encountered while parsing the response as JSON (provided by the browser)

**Comment by Haroenv** [comment]
There should be a scroll transition or a focus shift if you click on show files towards the first link so that it works nice on mobile

File: js/src/lib/Details/Aside.js (hunk 3 above)
Comment (Haroenv) [inline js/src/lib/Details/Aside.js]: That's fine. What about using the Unicode dot directly instead? No big deal anyway

**Comment by Haroenv** [comment]
I opened https://github.com/Daniel15/yarn-website/pull/1 on your fork for the icon change

File: _sass/_details.scss [hunk 7]
```diff
@@ -118,6 +118,58 @@
...
+    }
+    &__filename::before {
+      background-image: url(/assets/detail/ico-file.svg);
+    }
+
+    // Expand animation for subdirectories
+    &-enter {
+      max-height: 0;
+      opacity: 0;
+      transition: all 500ms ease-in;
```
Comment (Haroenv) [inline _sass/_details.scss]: I find this a bit long actually.

File: js/src/lib/Details/Aside.js (hunk 3 above)
Comment (Daniel15) [inline js/src/lib/Details/Aside.js]: Good point, I can do that. Internally at FB we have a lint rule blocking all non-ASCII characters for some reason. I could directly embed it here though, you're right :D

File: _sass/_details.scss (hunk 7 above)
Comment (Daniel15) [inline _sass/_details.scss]: Yeah... I'll bump it down a bit.

**Comment by Daniel15** [comment]
> There should be a scroll transition or a focus shift if you click on show files towards the first link so that it works nice on mobile

Good catch, I didn't even test on mobile.

File: js/src/lib/Details/FileBrowser.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+  render() {
+    return (
+      <div>
+        <h2 className="m-2">
+          {window.i18n.detail.files_header.replace(
+            '{name}',
+            this.props.objectID
+          )}
+        </h2>
+        <a href="#" onClick={this.props.onBackToDetails}>
```
Comment (Haroenv) [inline js/src/lib/Details/FileBrowser.js]: Maybe give this a ref to focus it on `componentWillMount` or similar

**Comment by Daniel15** [comment]
I just did a `.focus()` on mount to scroll to the right place on mobile. It might better to have a fancy animation, but this is good enough for now.

The other thing I wanted to do is have a different deep link for when viewing the file list (like `/package/react/browse-files` or something) but this is fine for now.

**Comment by Haroenv** [comment]
Yes, I was thinking about a different deep link too, with history.pushState() and checking the url on load, but it’s great like this already 😄

**Comment by Daniel15** [comment]
I'm going to merge this and we can iterate on it to make improvements as needed :)

**Pull request merged**